src/c-writer.cc

```diff
@@ -1790,15 +1802,50 @@ void CWriter::WriteCallIndirectFuncDeclaration(const FuncDeclaration& decl,
   Write(")");
 }
 
-void CWriter::WriteFeatureMacros() {
+static bool func_uses_simd(const FuncSignature& sig) {
+  return std::any_of(sig.param_types.begin(), sig.param_types.end(),
+                     [](auto x) { return x == Type::V128; }) ||
+         std::any_of(sig.result_types.begin(), sig.result_types.end(),
+                     [](auto x) { return x == Type::V128; });
+}
+
+void CWriter::ComputeSimdScope() {
+  simd_used_in_header_ =
+      module_->features_used.simd &&
+      (std::any_of(module_->globals.begin(), module_->globals.end(),
+                   [](const auto& x) { return x->type == Type::V128; }) ||
+       std::any_of(module_->imports.begin(), module_->imports.end(),
+                   [](const auto& x) {
+                     return x->kind() == ExternalKind::Func &&
+                            func_uses_simd(cast<FuncImport>(x)->func.decl.sig);
+                   }) ||
+       std::any_of(module_->exports.begin(), module_->exports.end(),
+                   [&](const auto& x) {
+                     return x->kind == ExternalKind::Func &&
+                            func_uses_simd(module_->GetFunc(x->var)->decl.sig);
+                   }));
+}
+
+void CWriter::WriteHeaderIncludes() {
+  Write("#include \"wasm-rt.h\"", Newline());
+
   if (module_->features_used.exceptions) {
-    Write("#define WASM_RT_ENABLE_EXCEPTION_HANDLING", Newline(), Newline());
+    Write("#include \"wasm-rt-exceptions.h\"", Newline(), Newline());
   }
-  if (module_->features_used.simd) {
-    Write("#define WASM_RT_ENABLE_SIMD", Newline(), Newline());
+
+  if (simd_used_in_header_) {
+    WriteV128Decl();
   }
 }
 
+void CWriter::WriteV128Decl() {
+  Write("#include <simde/wasm/simd128.h>", Newline(), Newline());
+  Write("#ifndef WASM_RT_SIMD_TYPE_DEFINED", Newline(),
```

Comment:
Do you need the guard given that you also have simd_used_in_header_?

Reply:
Unfortunately I think we still need the guard, given that this is sometimes going in the .h file, and C forbids duplicate typedefs even if they're identical. If the embedder includes two module .h files, we don't want a collision. :-/